micro-rdk/src/esp32/log.rs

```diff
@@ -0,0 +1,137 @@
+//! This is for infrastructure supporting uploading logs from an ESP32 to Viam's cloud infrastructure
+//! (see common/log.rs for more details). On an ESP32, there are two sources of logs: logs from ESP-IDF,
+//! and logs from various Rust crates. The EspLogger from esp-idf-svc efficiently redirects the latter
+//! logs to UART in a format matching the first, so it is sufficient to simply wrap it. However
+//! to capture ESP-IDF logs, it is necessary to replace the default vprintf function to write to LOG_BUFFER
+//! (again, see common/log.rs) before invoking the previously existing vprintf function in order to write to
+//! UART. We store the previous vprintf function in PREVIOUS_LOGGER and use esp_log_set_vprintf for this purpose.
+use crate::{
+    common::log::ViamLogEntry,
+    google::protobuf::{value::Kind, Struct, Value},
+    proto::common::v1::LogEntry,
+};
+
+use esp_idf_svc::log::EspLogger;
+use esp_idf_svc::sys::{esp_log_set_vprintf, va_list, vprintf_like_t};
+use printf_compat::output::display;
+use ringbuf::Rb;
+use std::{collections::HashMap, sync::OnceLock};
+use std::{ffi::c_char, sync::Mutex};
+
+use crate::common::log::{get_log_buffer, ViamLogAdapter};
+
+static PREVIOUS_LOGGER: OnceLock<Mutex<vprintf_like_t>> = OnceLock::new();
+static CURRENT_LOG_STATEMENT: OnceLock<Mutex<Vec<String>>> = OnceLock::new();
+
+const MESSAGE_START: &str = "\x1b[0;";
+const MESSAGE_END: &str = "\x1b[0m";
+
+// A single log statement is often broken up into multiple calls to vprintf. So we store
+// the fragments in CURRENT_LOG_STATEMENT. Detecting whether we have encountered the start
+// of a new statement is further complicated by the fact that, depending on the ESP-IDF component
+// producing the log, the statement can be in one of the following formats
+//
+// "\x1b[0;<color_indicator>m<level_indicator> ... \x1b[0m"
+//
+// "<level_indicator> (<timestamp>) ..."
+// This complication is reflected in the function below and its helper function
+// process_current_statement_and_level
+#[allow(improper_ctypes_definitions)]
+unsafe extern "C" fn log_handler(arg1: *const c_char, arg2: va_list) -> i32 {
+    let va_list: core::ffi::VaList = std::mem::transmute(&arg2);
+    let fmt_message = display(arg1, va_list);
+    let message = format!("{}", fmt_message).trim().to_string();
+    let message_clone = message.clone();
+    let start_of_new_statement = (message_clone.len() >= 3)
+        && (matches!(&message_clone[..3], "I (" | "E (" | "W (" | "D (" | "V (")
+            || message_clone.starts_with(MESSAGE_START));
+    let mut current_fragments = CURRENT_LOG_STATEMENT
+        .get_or_init(|| Mutex::new(vec![]))
+        .lock()
+        .unwrap();
+    if start_of_new_statement && !current_fragments.is_empty() {
+        let full_message = current_fragments.join(" ");
+        let _ = get_log_buffer()
+            .lock_blocking()
```

Comment:
this a bit risky IMO. You are attempting to acquire an AsyncLock outside and async context from C. If the caller thread has a higher prio it may lead to a deadlock.

Reply:
I believe this risk to be mitigated by the fact that this buffer is not publicly accessible, which reduces the chance that someone would acquire the lock and then attempt to log something before releasing it. All of the risk is contained within this crate